torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,763 @@
+                break
+
+    train_step_graph.recompile()
+    train_step_graph.print_readable()
+
+    # next we forward propagate to mark all the sharding
+    for node in train_step_graph.graph.nodes:
+        if node.op == "placeholder":
+            if placeholder_idx < num_params:
+                shard_sig = _gen_shard_strategy(mesh, 0)
+                replica_sig = _gen_replicate_strategy(mesh)
+                dp_strategy_map[node] = DataParallelStrategy(
+                    NodeType.PARAM, [replica_sig, shard_sig]
+                )
+
+            elif placeholder_idx < activation_idx:
+                # optimizer states follow the same strategy as
+                # the corresponding parameters
+                # NOTE: optimizer states can be None sometimes,
+                # creating a NON_TENSOR stratgy for those cases.
+                if "val" not in node.meta:
+                    dp_strategy_map[node] = DataParallelStrategy(
+                        NodeType.NON_TENSOR, []
+                    )
+                else:
+                    shard_sig = _gen_shard_strategy(mesh, 0)
+                    replica_sig = _gen_replicate_strategy(mesh)
+                    dp_strategy_map[node] = DataParallelStrategy(
+                        NodeType.STATE, [replica_sig, shard_sig]
+                    )
+            else:
+                # NOTE: args might not only contain activation but also
+                # contain other non-tensor inputs, creating a NON_TENSOR
+                # strategy for those cases.
+                if "val" not in node.meta:
+                    dp_strategy_map[node] = DataParallelStrategy(
+                        NodeType.NON_TENSOR, []
+                    )
+                else:
+                    activation_batch_dim_size = node.meta["val"].shape[batch_dim]
+                    # find the first activation node and use its batch dim size
+                    if batch_dim_analzer.batch_dim_size == -1:
+                        batch_dim_analzer.init_batch_dim_size(activation_batch_dim_size)
+
+                    batch_dim_analzer.set_batch_dim(node, batch_dim)
+                    shard_sig = _gen_shard_strategy(mesh, batch_dim)
+                    dp_strategy_map[node] = DataParallelStrategy(
+                        NodeType.ACT, [shard_sig]
+                    )
+            placeholder_idx += 1
+        elif node.op == "call_function":
+            # Annotate node types for the computation graph
+            # Data Parallel node propagation logic:
+            # param (non-compute) -> out: param
+            # grad (non-compute before/after) -> out: grad
+            # state -> output: state
+            #
+            # param + activation (param must be replicate, act be sharded) -> out: activation
+            # param/state + grad (param/state/grad be the same spec) -> out: param/state
+            # param + state -> out: param
+
+            if node.target in non_compute_ops:
```

Comment:
at this point, i assume `tag_grad` nodes should all have been removed. If that's the case, then both usage of `non_compute_ops` don't expect to see `tag_grad` in there?